kep/real-load-aware-scheduling/real-load-aware-scheduling.md

```diff
@@ -0,0 +1,540 @@
+    3. Metrics Provider failure: The metrics provider can fail to serve requests for several reasons like OOM, unresponsive components, etc. In this case, every node’s metrics are unavailable, and we fall back to the best fit based on allocations. 
+    4. Addition of new nodes: It is easy to handle this if no pods have been scheduled on them, so utilization must be 0. If any recent new pods are running already then we can use the method above for prediction. If there are long-running pods, we can avoid this node until metrics are available.
+2. Incorrect metrics (due to misconfiguration, bugs, etc.): It is challenging to detect such metrics. For the first implementation, we would like to ignore this case by assuming metrics are correct. However, we plan to add metrics from multiple providers to corroborate data points and minimize the effect of incorrect metrics on scheduling in the future. 
+
+### **Load Watcher API**
+
+
+JSON Schema - see Appendix
+
+JSON payload example:
+
+
+```
+{
+  "timestamp": 1556987522,
+  "window": {
+    "duration" : "15m",
+    "start" : 1556984522,
+    "end"   : 1556985422
+  },
+  "source" : "InfluxDB",
+  "data" :
+    {
+      "node-1" : {
+        "metrics" : [
+          {
+            "name" : "host.cpu.utilisation",
+            "type" : "cpu",
+            "rollup" : "AVG",
+            "value"  : 20
+          },
+          {
+            "name" : "host.memory.utilisation",
+            "type" : "memory",
+            "rollup" : "STD",
+            "value"  : 5
+          }],
+        "tags" : [{}],
+        "metadata" : {
+          "dataCenter" : "data-center-1",
+          "pool" : "critical-apps"
+        }
+      },
+      "node-2" : {
+        "metrics" : [
+          {
+            "name" : "host.cpu.utilisation",
+            "type"  : "cpu",
+            "rollup" : "AVG",
+            "value"  : 20
+          },
+          {
+            "name" : "host.memory.utilisation",
+            "type" : "memory",
+            "rollup" : "STD",
+            "value"  : 5
+          }
+        ]
+      },
+      "metadata" : {
+        "dataCenter" : "data-center-2",
+        "pool" : "light-apps"
+      },
+      "tags" : [{}]
+    }
+}
+```
+
+
+**REST API**
+
+
+```
+GET /watcher
+Returns metrics for all nodes in the cluster
+
+
+GET /watcher/{hostname}
+Returns metrics for the hostname given
+
+
+200 OK response example given above
+404 if no metrics are present
+```
+
+
+
+### **PostBind Plugin**
+
+A custom plugin will be added at post bind extension point, which maintains a time ordered state of node → pod mappings for pods scheduled successfully in last 5 minutes. This state will be maintained across scheduling cycles and used from BestFitBinPack Score plugin. This state will help to predict utilisation based on allocations when metrics are bad, especially in case 1 above (Unavailability of metrics).
```

Comment:
It seems this act as a backup remediation method? If metrics are all good, we don't need it right?

If so, I'd suggest to get this implemented as a separate "controller" watching goroutine, as (1) it won't slow the whole scheduling cycle, and (2) it's able to watch Pod placements by external schedulers.

Reply:
This is not exactly a backup remediation. It is a critical component to predict correct utilisation. For example: Consider 3 pods scheduled together. For scheduling the 2nd and 3rd pods, we wouldn't have the correct utilisation values from the nodes because time to schedule (under normal circumstances) << metric-agent reporting interval (usually 1 minute). Without maintaining this state we will end up underestimating the actual utilisation values we get from load watcher.

Reply:
I did some benchmarking and this doesn't slow the scheduling (considerably), it adds one function call which doesn't do any significant work.